python/note_input_widget/widget.py

```diff
@@ -360,7 +445,39 @@ def _async_submit_note(self, sg, data):
                                           "tasks": note_tasks })
         
         self.__upload_thumbnail(sg_note_data, sg, data)
+        self.__upload_attachments(sg_note_data, sg, data)
         
+
+    def __upload_attachments(self, parent_entity, sg, data):
+        """
+        Uploads any generic file attachments to Shotgun, parenting
+        them to the Note entity.
+
+        :param parent_entity:   The Note entity to attach the files to in SG.
+        :param sg:              A Shotgun API handle.
+        :param data:            The data dict containing an "attachments" key
+                                housing a list of file paths to attach.
+        """
+        for file_path in data.get("attachments", []):
+            if os.path.exists(file_path):
+                self.__upload_file(file_path, parent_entity, sg)
+
+
+    def __upload_file(self, file_path, parent_entity, sg):
+        """
+        Uploads any generic file attachments to Shotgun, parenting
+        them to the Note entity.
+
+        :param file_path:       The path to the file to upload to SG.
+        :param parent_entity:   The Note entity to attach the files to in SG.
+        :param sg:              A Shotgun API handle.
+        """
+        self._bundle.log_debug(
+            "Uploading attachments (%s bytes)..." % os.path.getsize(file_path)
+        )
+        sg.upload(parent_entity["type"], parent_entity["id"], str(file_path))
+        self._bundle.log_debug("Upload complete!")
```

Comment:
Is this something that happens in the foreground?  Would it be better to run it in a bg task runner?  Ideally we would be adding some decorations to the attachments so it is clear which have been uploaded and which have yet to be uploaded.